### DISCUSSION TITLE
How to use blockSize of 25 in cases with pre-inlet?

### DISCUSSION THREAD

**Discussion by mojtabaAmiraslanpour** [UvaCsl/HemoCell#120] [General discussion]
Hello,

I've been struggling to find a way to cut down on computation time. The current implementation would keep the whole blocks of the rectangular bounding box and only put bounce-back BCs on all nodes other than the actual fluid domain (by setting blockSize = -1), but this isn't efficient for complex geometries.

I've tried different methods to get it to work just like the simple pipeflow case, where unneeded blocks are removed, but nothing seems to work until now.

If anyone has any ideas, I'd really appreciate the help—it could save me a lot of time and resources.

Thanks.

**Comment by gzavo** [comment]
Hi @mojtabaAmiraslanpour ,

It is inefficient in terms of resources and overall energy consumption of the simulation, but does not really influence the computational time. Due to numerical limitations you are often bound to use 25^3 smallest domains. So dropping all the "empty" domains would lead to fewer needed CPU cores, but still the same runtime.

If you want to go this route there are two options I can think of: 
1) there is a function in palabos (reparallelize) that recomputes the atomic block structure and creates a sparse distribution. In theory it should be possible to use it with HemoCell (not tested).
2) we have a preliminary load-balancing mechanism (look for uses of METIS in the code). This works if you have many more atomic blocks than CPU cores. It assigns the blocks to CPU cores in a way that is closer to homogeneous load.

As said, neither will yield lower wall-times, but fewer necessary computing resources. In order to improve performance the challenge is to compute cells in a distributed way (i.e., going below the 25^3 limit). This is tricky, because some material forces (volume conservation) requires the whole cell to reside on the same CPU. This is not impossible of course, but not implemented at the moment. It would be a great addition though.
